cmd/lotus-shed/state-stats.go

```diff
@@ -0,0 +1,534 @@
+			Chain: cs,
+		}
+
+		ts, err := lcli.LoadTipSet(ctx, cctx, tsr)
+		if err != nil {
+			return err
+		}
+
+		log.Infow("tipset", "parentstate", ts.ParentState())
+
+		if len(addrs) == 0 && cctx.Bool("all") {
+			var err error
+			addrs, err = sm.ListAllActors(ctx, ts)
+			if err != nil {
+				return err
+			}
+		}
+
+		numWorkers := cctx.Int("workers")
+
+		var wg sync.WaitGroup
+
+		errChan := make(chan error)
+		jobs := make(chan address.Address, numWorkers)
+		results := make(chan actorStats, numWorkers)
+
+		workersCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		worker := func(ctx context.Context, id int, wg *sync.WaitGroup, jobs <-chan address.Address, results chan<- actorStats) error {
+			completed := 0
+
+			defer wg.Done()
+			defer func() {
+				log.Infow("worker done", "id", id, "completed", completed)
+			}()
+
+			for {
+				select {
+				case addr, ok := <-jobs:
+					if !ok {
+						return nil
+					}
+
+					actor, err := sm.LoadActor(workersCtx, addr, ts)
+					if err != nil {
+						return err
+					}
+
+					dag, err := newCacheNodeGetter(merkledag.NewDAGService(blockservice.New(bs, offline.Exchange(bs))), 8096)
+					if err != nil {
+						return err
+					}
+
+					actStats, err := collectStats(workersCtx, addr, actor, dag)
+					if err != nil {
+						return err
+					}
+
+					select {
+					case results <- actStats:
+					case <-ctx.Done():
+						return ctx.Err()
+					}
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+
+				completed = completed + 1
+			}
+		}
+
+		for w := 0; w < numWorkers; w++ {
+			wg.Add(1)
+			go func(id int) {
+				if err := worker(workersCtx, id, &wg, jobs, results); err != nil {
+					select {
+					case errChan <- err:
+					case <-workersCtx.Done():
+					}
+
+					return
+				}
+			}(w)
+		}
+
+		go func() {
+			wg.Wait()
+			close(results)
+		}()
+
+		go func() {
+			for _, addr := range addrs {
+				jobs <- addr
+			}
+
+			close(jobs)
+		}()
+
+		for {
+			select {
+			case result, ok := <-results:
+				if !ok {
+					return nil
+				}
+
+				if cctx.Bool("pretty") {
+					DumpStats(result)
+				} else {
+					if err := DumpJSON(result); err != nil {
+						return err
+					}
+				}
+			case err := <-errChan:
+				return err
+			case <-ctx.Done():
+				return ctx.Err()
+
+			}
+		}
+	},
+}
+
+func collectStats(ctx context.Context, addr address.Address, actor *types.Actor, dag format.NodeGetter) (actorStats, error) {
+	log.Infow("actor", "addr", addr, "code", actor.Code, "name", builtin.ActorNameByCode(actor.Code))
+
+	nd, err := dag.Get(ctx, actor.Head)
+	if err != nil {
+		return actorStats{}, err
+	}
+
+	// When it comes to fvm / evm actors this method of inspecting fields will probably not work
+	// and we may only be able to collect stats for the top level object. We might be able to iterate
+	// over the top level fields for the actors and identify field that are CIDs, but unsure if we would
+	// be able to identify a field name.
```

Comment:
You are right about this for M2.2, we'd need some open manifest so that something like DumpActorState gets updated with meaningful field names. For FeVM we use a single builtin actor type to manage all user contracts so this will actually work pretty well, though we'll have the same problem one level down on the user contract code hash.